Implement client-side pre- and post-commit hooks

RLThomaz opened the pull request:
Implement client-side pre- and post-commit hooks JohnstonCode/svn-scm#1355

I've tested it locally and it works using VSCode commands, but did not work directly on the SCM - is it supposed to in debug mode?

RLThomaz committed:
Implement client-side pre- and post-commit hooks JohnstonCode/svn-scm#1355

RLThomaz committed:
Merge branch 'master' of https://github.com/JohnstonCode/svn-scm

RLThomaz committed:
Merge branch 'master' into master

JohnstonCode commented:
Do you think there would be more benefit from setting the hooks in the config to be an array of string path so multiple hooks could be added for each hook?

JohnstonCode commented:
Your current implementation will only work if you run the commit command. I'm not 100% sure where the best place would be to put the hook check code.

RLThomaz commented:
> Do you think there would be more benefit from setting the hooks in the config to be an array of string path so multiple hooks could be added for each hook?

Sure, I cannot think of any reason against it. Actually, I believe this would be of great benefit for multi-root workspaces. I'm thinking of sending the repository name as an input argument to the hooks. This allows the user to control if a hook should do anything for a given repository. SVN does something similar but includes the transaction id - which we won't have on the client side.

> Your current implementation will only work if you run the commit command. I'm not 100% sure where the best place would be to put the hook check code.

Yeah, that's what I was talking about in the PR comment. I'm not - at all - familiar with VSCode extensions, I understand that you can register commands - triggered manually using the command palette `Ctrl + Shift + P` - and that's the current behavior. However, how does the SCM invoke this extension? Shouldn't it just call the command for us? I've noticed that using the SCM still triggers the extension, so which file is it actually calling?

RLThomaz committed:
Improve client-side commit hooks JohnstonCode/svn-scm#1355
 - Hooks are now their own class containing information such as project and branch.
 - Settings were updated to reflect the new Hook class.

RLThomaz commented:
@JohnstonCode, I've just committed an improved version. Now the settings allow an array of objects, each object contains an array/string of commands. In addition, each object may contain a project name and branch. That way the hooks are only executed for the corresponding repository/project/branch. If project/branch is omitted the hooks will execute for all/any options available.

I've noticed I'm not really solving the problem in #1355, they want to be able to change the actual commit-transaction body. I think that's possible, but would require very specific hooks regarding the scope of what they do - in this case, the command would have to return back the updated transaction body. What do you think?

JohnstonCode commented:
I think the hooks would be better placed here https://github.com/JohnstonCode/svn-scm/blob/7b45ad36bb21ebb71f72e140664b12cdf8b6f329/src/repository.ts#L1056 As your current hook only run on the single commit commands and not the commitWithMessage command

RLThomaz commented:
> I think the hooks would be better placed here
> 
> https://github.com/JohnstonCode/svn-scm/blob/7b45ad36bb21ebb71f72e140664b12cdf8b6f329/src/repository.ts#L1056

Interesting, so there is an enumeration specifying which command is being executed, correct? Ok, I can totally see the benefits of running the hooks there, but it would be just a call to a public function of `Hook` passing as arguments the `operation (Operation)` and `this (Repository)`. We would need two functions in `Hook`, a pre-operation call and a post-operation call. This way those functions would be responsible for figuring out which hooks should be called given the operation type.

> As your current hook only run on the single commit commands and not the commitWithMessage command

By the way, it's there on [`commitWithMessage`](https://github.com/JohnstonCode/svn-scm/pull/1434/files#diff-12a1e225f8f6a70fe6b1ae5bfb0f4ae6dbfc1906c79e8fdbb6f234bf54e0ef8d) too.

RLThomaz committed:
Move client-side hook execution to repository general run function JohnstonCode/svn-scm#1355
 - Hooks are now more general, including type, disposition, and operation.
 - Hooks can now be executed for any SVN operation.
 - Hooks can execute system calls or VSCode commands.

RLThomaz commented:
Alright, just pushed a more generic implementation of `Hook`.

 - I followed your suggestion and moved the execution to https://github.com/JohnstonCode/svn-scm/blob/7b45ad36bb21ebb71f72e140664b12cdf8b6f329/src/repository.ts#L1056
 - Hooks now hold several properties.
  - Disposition: pre or post-command execution.
  - Command Type: System call or VS Code command.
  - Operation: SVN operation type on which the commands should be executed.

I believe the addition of VS Code command support should give us something similar to @rroman6174's [implementation](https://github.com/rroman6174/svn-scm). The only thing that's lacking now is a solution to #1355. Perhaps we can use `stdout` [from here](https://github.com/RLThomaz/svn-scm/blob/master/src/helpers/hooks.ts#L50) to update whatever the command needs - for example, the commit message. What do you think, @JohnstonCode?

Here's one example of what we can do now:

```json
{
"svn.hooks": [
    {
      "commandDisposition": "Pre",
      "commandType": "VSCode",
      "commandOperation": "Commit",
      "command": "search.action.openNewEditor",
      "branch": "trunk",
      "project": "MyProject"
    },
    {
      "commandDisposition": "Pre",
      "commandType": "System",
      "commandOperation": "Commit",
      "command": "/bin/bash -c 'echo \"Pre-commit hook failed: \n this is an error.\" 1>&2; exit 1' ",
      "branch": "trunk",
      "project": "MyProject"
    }
  ]
}
```

RLThomaz committed:
Update README.md

JohnstonCode commented on README.md:
```diff
@@ -118,6 +118,9 @@ Here are all of the extension settings with their default values. To change any
   "svn.gravatars.enabled": true,
 
+  // Set up client-side hooks to be executed before or after SVN commands.
+  "svn.hooks": null
```
This should just be an empty array otherwise you will need to add null checks

RLThomaz replied:
Good catch! I'll fix this soon along with other issues I have identified.

RLThomaz committed:
Implement overwritable SVN operation arguments JohnstonCode/svn-scm#1355
 - This is just a draft and only let `commit` operations overwrite SVN arguments.
 - Fix null hooks configuration.

RLThomaz commented on src/repository.ts:
```diff
@@ -842,8 +842,12 @@ export class Repository implements IRemoteRepository {
 
   public async commitFiles(message: string, files: any[]) {
-    return this.run(Operation.Commit, () =>
-      this.repository.commitFiles(message, files)
+    const args = { message: message, files: files };
+    return this.run(
+      Operation.Commit,
+      (obj: any = { message: message, files: files }) =>
+        this.repository.commitFiles(obj.message, obj.files),
+      args
     );
   }
```
This is an example of how we are supposed to allow operations to have arguments overwritten by pre-operation hooks. Instead of an empty arrow function, we create one taking an object as an argument. This object holds a reference to the original arguments, but allow a new object to be passed overwriting the original arguments.

RLThomaz committed:
Merge branch 'JohnstonCode:master' into master

RLThomaz committed:
Merge branch 'master' into master

JohnstonCode committed:
Merge branch 'master' into master
